File: leetcode/Leet1209.java
package me.algo.leetcode;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Created by Bomi on 2020/10/21.
 * 문제 출처 : https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string-ii/
 *
 * Time Complexity : O(N)
 * Used Algorithm : String
 * Used Data structure : Deque(ArrayDeque)
 */

public class Leet1209 {
    static class Pair {
        char ch;
        int count;

        Pair(char ch, int count) {
            this.ch = ch;
            this.count = count;
        }
    }

    public String removeDuplicates(String s, int k) {
        Deque<Pair> queue = new ArrayDeque<>();
        for (char c : s.toCharArray()) {
            if (queue.isEmpty() || queue.peekLast().ch != c) {
                queue.offerLast(new Pair(c, 1));
                continue;
            }
            queue.peekLast().count++;
            while (!queue.isEmpty() && queue.peekLast().count == k) {
                queue.pollLast();
            }
        }

        StringBuilder sb = new StringBuilder();
        while (!queue.isEmpty()) {
            Pair pair = queue.pollFirst();
            for (int i = 0; i < pair.count; i++) {
                sb.append(pair.ch);
            }
        }
        return sb.toString();
    }
}

File: week54/sun/Leet1209.kt
package me.algo.week54.sun

import java.util.*

/**
 * Created by Bomi on 2021/05/18.
 * 문제 출처 : https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string-ii/
 *
 * Time Complexity : O(N)
 * Used Algorithm : Stack
 * Used Data structure : Stack
 */

class Leet1209 {
    fun removeDuplicates(s: String, k: Int): String {
        val stack = Stack<Pair<Char, Int>>()
        s.forEach {
            if (stack.isEmpty() || stack.peek().first != it) {
                stack.push(Pair(it, 1))
            } else {
                stack.push(Pair(it, stack.pop().second + 1))
            }
            if (stack.peek().second == k) {
                stack.pop()
            }
        }

        fun toJoinString(): String {
            var result = ""
            stack.forEach { p ->
                repeat(p.second) {
                    result += p.first
                }
            }
            return result
        }
        return toJoinString()
    }
}